experimental/Schemes/Resolution_structure.jl

```diff
@@ -0,0 +1,738 @@
+    R = ambient_coordinate_ring(U)
+    JM = jacobian_matrix(R, mod_gens)
+    if length(amb_col) < length(mod_gens)
+      JM_essential = JM[:, amb_col]
+    else
+      JM_essential = JM
+    end
+    submat_for_minor = JM[amb_row, amb_col]
+    Ainv, h2 = pseudo_inv(submat_for_minor)
+    h == h2 || error("inconsistent input data")
+    JM_essential = JM_essential * Ainv
+    I_gens = lifted_numerator.(gens(I))
+    JI = jacobian_matrix(I_gens)
+    result_mat = h*JI
+    for i in 1:length(amb_col)
+      for j in 1:ncols(result_mat)
+        result_mat[:,j] = result_mat[:,j] - [JM_essential[k,i] * JI[amb_row[i],j] for k in 1:nrows(JM_essential)]
+      end
+    end
+    Ivec = copy(gens(I))
+    append!(Ivec,[a for a in OO(U).(collect(result_mat))])
+    Delta_dict[U] = ideal(OO(U),Ivec)
+  end
+
+  return small_generating_set(IdealSheaf(W,Delta_dict))
+end
+ 
+########################################################################
+# test for snc                                                         #
+########################################################################
+
+function divisor_intersections_with_X(current_div, I_X)
+  scheme(I_X) == scheme(current_div[1]) || error("underlying schemes do not match")
+  n_max = dim(I_X)
+
+  inter_div_dict = Dict{Vector{Int},Tuple{AbsIdealSheaf,Int}}()
+  old_keys = Vector{Int}[]
+  empty_keys = Vector{Int}[]
+  essential_inter = AbsIdealSheaf[]
+
+# initialization: each divisor + I_X
+  for k in 1:length(current_div)
+    Idiv = ideal_sheaf(current_div[k]) + I_X
+    if !is_one(Idiv)
+      inter_div_dict[[k]] = (Idiv,0)
+      push!(old_keys, [k])
+    end
+  end
+  new_keys = copy(empty_keys)
+
+# add intersections
+  while !is_empty(old_keys)
+    old_keyvec = popfirst!(old_keys)         # this is the intersection to which we add a new divisor
+    for i in (old_keyvec[end]+1):length(current_div)
+      copykey = copy(old_keyvec)
+      if haskey(inter_div_dict, [i])
+        push!(copykey,i)                     # here we add it
+        Idiv = inter_div_dict[[i]][1] + inter_div_dict[old_keyvec][1]
+        if !is_one(Idiv)
+                                             # it intersects
+          subsetlist = subsets(old_keyvec,length(old_keyvec)-1)
+          subsetlist = [push!(a,i) for a in subsetlist]
+          if (sum([inter_div_dict[a][2] for a in subsetlist])> 0)
+                                             # offending intersection, known before
+            inter_div_dict[copykey] = (Idiv,2)
+            push!(new_keys,copykey)
+          elseif dim(Idiv) == n_max - length(copykey)
+                                             # non-offending intersection
+            inter_div_dict[copykey] = (Idiv,0)
+            push!(new_keys,copykey)
+          else
+                                             # offending intersection, new
+            inter_div_dict[copykey] = (Idiv,1)
+            push!(new_keys,copykey)
+            push!(essential_inter, Idiv)
+          end
```

Comment:
Extract common code?
```suggestion
          if (sum([inter_div_dict[a][2] for a in subsetlist])> 0)
                                             # offending intersection, known before
            inter_div_dict[copykey] = (Idiv,2)
          elseif dim(Idiv) == n_max - length(copykey)
                                             # non-offending intersection
            inter_div_dict[copykey] = (Idiv,0)
          else
                                             # offending intersection, new
            inter_div_dict[copykey] = (Idiv,1)
            push!(essential_inter, Idiv)
          end
          push!(new_keys, copykey)
```

Reply:
done